src/include/storage/write_ahead_log/log_manager.h

```diff
@@ -0,0 +1,105 @@
+
+  /**
+   * Must be called when no other threads are doing work
+   */
+  void Shutdown() {
+    Process();
+    Flush();
+    out_.Close();
+  }
+
+  /**
+   * Returns a (perhaps partially) filled log buffer to the log manager to be consumed. Caller should drop its
+   * reference to the buffer after the method returns immediately, as it would no longer be safe to read from or
+   * write to the buffer. This method can be called safely from concurrent execution threads.
+   *
+   * @param buffer the (perhaps partially) filled log buffer ready to be consumed
+   */
+  void AddBufferToFlushQueue(BufferSegment *buffer) { flush_queue_.Enqueue(buffer); }
+
+  /**
+   * Register a callback for the committed transaction beginning at the given time, such that the callback will be
+   * called as soon as possible by the LogManager when its commit record is out. Behavior is not defined if the provided
+   * transaction is aborted or uncommitted and the callback might never be invoked.
+   *
+   * @param txn_begin begin timestamp of the transaction to observe on
+   * @param callback the callback to invoke when the commit record is persistent; this callback will be invoked on the
+   *                 serializing thread, and should be kept short and fast, performing no expensive computation or waits
+   *                 on resources.
+   */
+  void RegisterTransactionFlushedCallback(timestamp_t txn_begin, const std::function<void()> &callback) {
+    common::SpinLatch::ScopedSpinLatch guard(&callbacks_latch_);
+    auto ret UNUSED_ATTRIBUTE = callbacks_.emplace(txn_begin, callback);
+    TERRIER_ASSERT(ret.second, "Insertion failed, callback is already registered for given transaction");
+  }
+
+  /**
+   * Process all the accumulated log records and serialize them out to disk. Flush can happen immediately or later
+   * depending on the state of the LogManager, and an explicit call to Flush() is required for any guarantee. (Beware
+   * the performance consequences of calling flush too frequently) This method should only be called from a dedicated
+   * logging thread.
+   */
+  void Process();
+
+  /**
+   * Flush the logs to make sure all serialized records before this invocation are persistent. Callbacks from committed
+   * transactions are also invoked when possible. This method should only be called from a dedicated logging thread.
+   */
+  void Flush();
+
+ private:
+  // TODO(Tianyu): This can be changed later to be include things that are not necessarily backed by a disk
+  // (e.g. logs can be streamed out to the network for remote replication)
+  BufferedLogWriter out_;
+  RecordBufferSegmentPool *buffer_pool_;
+
+  // These need to be thread-safe since various execution threads will modify these
+  common::ConcurrentQueue<BufferSegment *> flush_queue_;
+
+  // TODO(Tianyu): Might not be necessary, since commit on txn manager is already protected with a latch
+  common::SpinLatch callbacks_latch_;
```

Comment:
Mixed usage of concurrent data structure and latches. Is it possible to use just one kind?

Reply:
I will use only a latch for now and add a TODO for later